ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandStore.java

```diff
@@ -55,6 +62,7 @@
   private final Producer<CommandId, Command> commandProducer;
   private final CommandIdAssigner commandIdAssigner;
   private final Map<CommandId, QueuedCommandStatus> commandStatusMap;
+  private List<Pair<CompletableFuture<Void>, Long>> commandOffsetFutures;
```

Comment:
What threads are accessing `commandOffsetFutures`?  Only one? Or multiple? Cos `ArrayList` isn't thread-safe...

Also, generally try and avoid `Pair`.  All that `getLeft()` `getRight()` is ugly and makes the code harder to read.  Doesn't take much time to knock out a quick:

```
private static final class OffsetFuture  {
    final CompleteableFuture<Void> future = new ...;
    final long offset;
    
    private OffsetFuture(offset) {
    ...}
}

Which will make your code much more readable.

Having said that. I think in this instance a `Map<Offset, CompleteableFuture<Void>>` might be better.  Why?   Well ... 

- You can use ConcurrentMap impls if you need.
- You can use a TreeMap to only iterate only over the set of offets that need completing.
- You can share futures if multiple requests want the same offset.
- It removes the `Pair` which is good ;)